mlos_bench/mlos_bench/tests/optimizers/grid_search_optimizer_test.py

```diff
@@ -0,0 +1,246 @@
+    """
+    return {
+        "grid": {
+            "cost": 1,
+            "params": {
+                "cat": {
+                    "type": "categorical",
+                    "values": ["a", "b", "c"],
+                    "default": "a",
+                },
+                "int": {
+                    "type": "int",
+                    "range": [1, 3],
+                    "default": 2,
+                },
+                "float": {
+                    "type": "float",
+                    "range": [0, 1],
+                    "default": 0.5,
+                    "quantization": 0.25,
+                },
+            },
+        },
+    }
+
+
+@pytest.fixture
+def grid_search_tunables_grid(grid_search_tunables: TunableGroups) -> List[Dict[str, TunableValue]]:
+    """
+    Test fixture for grid from tunable groups.
+    Used to check that the grids are the same (ignoring order).
+    """
+    tunables_params_values = [tunable.values for tunable, _group in grid_search_tunables if tunable.values is not None]
+    tunable_names = tuple(tunable.name for tunable, _group in grid_search_tunables if tunable.values is not None)
+    return list(dict(zip(tunable_names, combo)) for combo in itertools.product(*tunables_params_values))
+
+
+@pytest.fixture
+def grid_search_tunables(grid_search_tunables_config: dict) -> TunableGroups:
+    """
+    Test fixture for grid search optimizer tunables.
+    """
+    return TunableGroups(grid_search_tunables_config)
+
+
+@pytest.fixture
+def grid_search_opt(grid_search_tunables: TunableGroups,
+                    grid_search_tunables_grid: List[Dict[str, TunableValue]]) -> GridSearchOptimizer:
+    """
+    Test fixture for grid search optimizer.
+    """
+    assert len(grid_search_tunables) == 3
+    # Test the convergence logic by controlling the number of iterations to be not a
+    # multiple of the number of elements in the grid.
+    max_iterations = len(grid_search_tunables_grid) * 2 - 3
+    return GridSearchOptimizer(tunables=grid_search_tunables, config={
+        "max_iterations": max_iterations,
+        "optimization_direction": "max",
+    })
+
+
+def test_grid_search_grid(grid_search_opt: GridSearchOptimizer,
+                          grid_search_tunables: TunableGroups,
+                          grid_search_tunables_grid: List[Dict[str, TunableValue]]) -> None:
+    """
+    Make sure that grid search optimizer initializes and works correctly.
+    """
+    # Check the size.
+    expected_grid_size = math.prod(tunable.cardinality for tunable, _group in grid_search_tunables)
+    assert expected_grid_size > len(grid_search_tunables)
+    assert len(grid_search_tunables_grid) == expected_grid_size
+    # Check for specific example configs inclusion.
+    expected_config_example: Dict[str, TunableValue] = {
+        "cat": "a",
+        "int": 2,
+        "float": 0.75,
+    }
+    grid_search_opt_pending_configs = list(grid_search_opt.pending_configs)
```

Comment:
```suggestion
    grid_search_opt_pending_configs = set(grid_search_opt.pending_configs)
```
if we want to check for subsets/supersets